File: lib/src/parser/score_parser.dart
import 'package:musicxml_parser/src/exceptions/musicxml_structure_exception.dart';
import 'package:musicxml_parser/src/models/appearance.dart';
import 'package:musicxml_parser/src/models/credit.dart'; // Import for Credit
import 'package:musicxml_parser/src/models/identification.dart';
import 'package:musicxml_parser/src/models/page_layout.dart';
import 'package:musicxml_parser/src/models/score.dart';
import 'package:musicxml_parser/src/models/staff_layout.dart'; // New import
import 'package:musicxml_parser/src/models/system_layout.dart'; // New import
import 'package:musicxml_parser/src/models/work.dart';
import 'package:musicxml_parser/src/parser/page_layout_parser.dart'; // New import
import 'package:musicxml_parser/src/parser/part_parser.dart';
import 'package:musicxml_parser/src/parser/staff_layout_parser.dart'; // New import
import 'package:musicxml_parser/src/parser/system_layout_parser.dart'; // New import
import 'package:musicxml_parser/src/parser/xml_helper.dart';
import 'package:musicxml_parser/src/utils/warning_system.dart';
import 'package:xml/xml.dart';

/// Parser for MusicXML score elements.
class ScoreParser {
  /// The parser for part elements.
  final PartParser _partParser;

  /// The warning system for collecting non-critical issues.
  final WarningSystem warningSystem;

  /// Creates a new [ScoreParser].
  ///
  /// [partParser] - Optional part parser. If not provided, a new one will be created.
  /// [warningSystem] - Optional warning system. If not provided, a new one will be created.
  ScoreParser({
    PartParser? partParser,
    WarningSystem? warningSystem,
  })  : warningSystem = warningSystem ?? WarningSystem(),
        _partParser = partParser ??
            PartParser(warningSystem: warningSystem ?? WarningSystem());

  /// Parses a MusicXML document into a [Score] object.
  ///
  /// [document] - The XML document to parse.
  Score parse(XmlDocument document) {
    // Check for score-partwise format (most common)
    final scorePartwise = document.findElements('score-partwise').firstOrNull;
    if (scorePartwise != null) {
      return _parseScorePartwise(scorePartwise);
    }

    // Check for score-timewise format
    final scoreTimewise = document.findElements('score-timewise').firstOrNull;
    if (scoreTimewise != null) {
      throw MusicXmlStructureException(
        'Score-timewise format is not fully implemented',
        requiredElement: 'score-partwise',
        parentElement: 'score-timewise',
        line: XmlHelper.getLineNumber(scoreTimewise),
      );
    }

    // Invalid MusicXML document
    throw MusicXmlStructureException(
      'Document is not a valid MusicXML file. Root element must be either "score-partwise" or "score-timewise"',
      requiredElement: 'score-partwise or score-timewise',
      line: XmlHelper.getLineNumber(document.rootElement),
    );
  }

  /// Parses a score-partwise element into a [Score] object.
  Score _parseScorePartwise(XmlElement element) {
    // Extract score metadata
    final title =
        XmlHelper.findOptionalTextElement(element, 'work/work-title') ??
            XmlHelper.findOptionalTextElement(element, 'movement-title');
    final composer = XmlHelper.findOptionalTextElement(
        element, 'identification/creator[@type="composer"]');
    final version = element.getAttribute('version');

    // Find part-list
    final partList = element.findElements('part-list').firstOrNull;

    if (partList == null) {
      warningSystem.addWarning(
        'Missing part-list element in score',
        category: 'structure',
        context: {
          'line': XmlHelper.getLineNumber(element),
        },
      );
    }

    // Parse parts
    final parts = element
        .findElements('part')
        .map(
          (part) => _partParser.parse(part, partList),
        )
        .toList();

    // Parse defaults section
    final defaults = _parseDefaults(element.getElement('defaults'));

    // Initialize ScoreBuilder
    final scoreBuilder = ScoreBuilder(version: version, line: XmlHelper.getLineNumber(element))
        .setTitle(title)
        .setComposer(composer);

    // Set parts
    scoreBuilder.setParts(parts);

    // Set defaults
    scoreBuilder
        .setPageLayout(defaults.pageLayout)
        .setDefaultSystemLayout(defaults.systemLayout)
        .setDefaultStaffLayouts(defaults.staffLayouts)
        .setScaling(defaults.scaling)
        .setAppearance(defaults.appearance);

    // Set work and identification
    if (title != null) {
      scoreBuilder.setWork(Work(title: title));
    }
    if (composer != null) {
      scoreBuilder.setIdentification(Identification(composer: composer));
    }

    // Set credits
    final parsedCredits = _parseCredits(element);
    if (parsedCredits.isNotEmpty) {
      scoreBuilder.setCredits(parsedCredits);
    }

    return scoreBuilder.build();
  }

  /// Helper structure to hold parsed default values.
  _DefaultsData _parseDefaults(XmlElement? defaultsElement) {
    if (defaultsElement == null) {
      return _DefaultsData();
    }

    final scaling = defaultsElement.getElement('scaling') != null
        ? ScalingParser().parse(defaultsElement.getElement('scaling')!)
        : null;

    final pageLayout = defaultsElement.getElement('page-layout') != null
        ? PageLayoutParser().parse(defaultsElement.getElement('page-layout')!)
        : null;

    final systemLayout = defaultsElement.getElement('system-layout') != null
        ? SystemLayoutParser().parse(defaultsElement.getElement('system-layout')!)
        : null;

    final staffLayouts = defaultsElement
        .findElements('staff-layout')
        .map((el) => StaffLayoutParser().parse(el))
        .toList();

    final appearance = _parseAppearance(defaultsElement);

    return _DefaultsData(
      scaling: scaling,
      pageLayout: pageLayout,
      systemLayout: systemLayout,
      staffLayouts: staffLayouts,
      appearance: appearance,
    );
  }

  List<Credit> _parseCredits(XmlElement scoreElement) {
    List<Credit> parsedCredits = [];
    final creditElements = scoreElement.findElements('credit');
    for (final creditElement in creditElements) {
      String? pageStr = creditElement.getAttribute('page');
      int? page = (pageStr != null && pageStr.isNotEmpty ? int.tryParse(pageStr) : null);

      XmlElement? creditTypeElement = creditElement.findElements('credit-type').firstOrNull;
      String? creditType = creditTypeElement?.innerText.trim();

      List<String> creditWordsList = [];
      final creditWordsElements = creditElement.findElements('credit-words');
      for (final wordsElement in creditWordsElements) {
        final text = wordsElement.innerText.trim();
        if (text.isNotEmpty) {
            creditWordsList.add(text);
        }
      }

      if ((creditType != null && creditType.isNotEmpty) || creditWordsList.isNotEmpty) {
         parsedCredits.add(Credit(page: page, creditType: creditType, creditWords: creditWordsList));
      } else if (page != null) {
         parsedCredits.add(Credit(page: page, creditType: creditType, creditWords: creditWordsList));
      }
    }
    return parsedCredits;
  }

  /// Parses the appearance information from the defaults element.
  Appearance? _parseAppearance(XmlElement? defaultsElement) {
    if (defaultsElement == null) return null;
    final appearanceElement = defaultsElement.getElement('appearance');
    if (appearanceElement == null) return null;

    final lineWidths = <LineWidth>[];
    final noteSizes = <NoteSize>[];

    // Parse line widths
    for (final lineWidthElement
        in appearanceElement.findElements('line-width')) {
      final type = lineWidthElement.getAttribute('type');
      if (type == null) continue;

      final width = double.tryParse(lineWidthElement.innerText);
      if (width == null) continue;

      lineWidths.add(LineWidth(
        type: type,
        width: width,
      ));
    }

    // Parse note sizes
    for (final noteSizeElement in appearanceElement.findElements('note-size')) {
      final type = noteSizeElement.getAttribute('type');
      if (type == null) continue;

      final size = double.tryParse(noteSizeElement.innerText);
      if (size == null) continue;

      noteSizes.add(NoteSize(
        type: type,
        size: size,
      ));
    }

    return Appearance(
      lineWidths: lineWidths,
      noteSizes: noteSizes,
    );
  }
}

/// Helper class to store parsed data from the <defaults> element.
class _DefaultsData {
  final Scaling? scaling;
  final PageLayout? pageLayout;
  final SystemLayout? systemLayout;
  final List<StaffLayout> staffLayouts;
  final Appearance? appearance;

  _DefaultsData({
    this.scaling,
    this.pageLayout,
    this.systemLayout,
    this.staffLayouts = const [],
    this.appearance,
  });
}

File: MusicXMLParser/Parser/ScoreParser.cs
using System.Xml.Linq;
using MusicXMLParser.Models;
using MusicXMLParser.Exceptions;
using MusicXMLParser.Utils;
using System.Linq;
using System.Collections.Generic;

namespace MusicXMLParser.Parser
{
    public class ScoreParser
    {
        private readonly PartParser _partParser;
        private readonly ScalingParser _scalingParser;
        private readonly PageLayoutParser _pageLayoutParser;
        private readonly SystemLayoutParser _systemLayoutParser;
        private readonly StaffLayoutParser _staffLayoutParser; // Assuming this parser exists
        public WarningSystem WarningSystem { get; }

        // 缓存常用的上下文字典以减少内存分配
        private readonly Dictionary<string, object> _sharedContext = new();

        public ScoreParser(
            PartParser? partParser = null,
            ScalingParser? scalingParser = null,
            PageLayoutParser? pageLayoutParser = null,
            SystemLayoutParser? systemLayoutParser = null,
            StaffLayoutParser? staffLayoutParser = null,
            WarningSystem? warningSystem = null)
        {
            WarningSystem = warningSystem ?? new WarningSystem();
            _partParser = partParser ?? new PartParser(warningSystem: WarningSystem);
            _scalingParser = scalingParser ?? new ScalingParser();
            _pageLayoutParser = pageLayoutParser ?? new PageLayoutParser();
            _systemLayoutParser = systemLayoutParser ?? new SystemLayoutParser();
            _staffLayoutParser = staffLayoutParser ?? new StaffLayoutParser();
        }

        public Score Parse(XElement element)
        {
            var line = XmlHelper.GetLineNumber(element);
            var version = XmlHelper.GetAttributeValue(element, "version");

            if (string.IsNullOrEmpty(version))
            {
                throw new MusicXmlValidationException(
                    message: "Score element is missing required 'version' attribute",
                    line: line,
                    context: CreateContext()
                );
            }

            return ParseScorePartwise(element);
        }

        private Score ParseScorePartwise(XElement element)
        {
            var line = XmlHelper.GetLineNumber(element);
            var version = XmlHelper.GetAttributeValue(element, "version");

            var scoreBuilder = new ScoreBuilder(version ?? "3.0");

            // 使用优化的方法获取元素
            var workElement = element.Element("work");
            if (workElement != null)
            {
                scoreBuilder.SetWork(ParseWork(workElement));
            }

            var identificationElement = element.Element("identification");
            if (identificationElement != null)
            {
                scoreBuilder.SetIdentification(ParseIdentification(identificationElement));
            }

            var defaultElement = element.Element("defaults");
            if (defaultElement != null)
            {
                ParseDefaults(defaultElement, scoreBuilder);
            }

            // 处理所有<credit>元素，优先设置Title
            var creditElements = element.Elements("credit").ToList();
            bool titleSet = false;
            foreach (var creditElement in creditElements)
            {
                var creditType = XmlHelper.GetElementText(creditElement, "credit-type");
                if (!titleSet && creditType == "title")
                {
                    // 合并所有credit-words内容
                    var titleText = string.Join(" ", creditElement.Elements("credit-words").Select(e => e.Value.Trim()).Where(s => !string.IsNullOrEmpty(s)));
                    if (!string.IsNullOrEmpty(titleText))
                    {
                        scoreBuilder.SetTitle(titleText);
                        titleSet = true;
                    }
                }
                scoreBuilder.AddCredit(ParseCredit(creditElement));
            }

            var partListElement = element.Element("part-list");
            if (partListElement == null)
            {
                throw new MusicXmlStructureException(
                    message: "Score element is missing required <part-list> child",
                    requiredElement: "part-list",
                    parentElement: "score",
                    line: line,
                    context: CreateContext()
                );
            }

            // 调试输出 part-list 下的元素数量
            Console.WriteLine($"[ScoreParser] part-list children count: {partListElement.Elements().Count()}");

            var parts = ParsePartList(partListElement);
            scoreBuilder.SetParts(parts);

            return scoreBuilder.Build();
        }

        // 优化的上下文创建方法，减少内存分配
        private Dictionary<string, object> CreateContext(string? additionalKey = null, object? additionalValue = null)
        {
            _sharedContext.Clear();
            
            if (additionalKey != null && additionalValue != null)
            {
                _sharedContext[additionalKey] = additionalValue;
            }
            
            return new Dictionary<string, object>(_sharedContext);
        }

        private Work ParseWork(XElement workElement)
        {
            // 使用优化的方法获取元素文本
            var workTitle = XmlHelper.GetElementText(workElement, "work-title");

            return new Work(workTitle);
        }

        private Identification ParseIdentification(XElement identificationElement)
        {
            var creator = XmlHelper.GetElementText(identificationElement, "creator");
            var rights = XmlHelper.GetElementText(identificationElement, "rights");
            var source = XmlHelper.GetElementText(identificationElement, "source");
            var encoding = ParseEncoding(identificationElement.Element("encoding"));

            return new Identification(creator, null, null, rights, source, encoding);
        }

        private Encoding? ParseEncoding(XElement? encodingElement)
        {
            if (encodingElement == null) return null;

            var software = XmlHelper.GetElementText(encodingElement, "software");
            var encodingDate = XmlHelper.GetElementText(encodingElement, "encoding-date");
            var supports = XmlHelper.GetElementText(encodingElement, "supports");

            return new Encoding(software, encodingDate, supports);
        }

        private void ParseDefaults(XElement defaultsElement, ScoreBuilder scoreBuilder)
        {
            // 使用优化的方法获取元素
            var scalingElement = defaultsElement.Element("scaling");
            if (scalingElement != null)
            {
                scoreBuilder.SetScaling(_scalingParser.Parse(scalingElement));
            }

            var pageLayoutElement = defaultsElement.Element("page-layout");
            if (pageLayoutElement != null)
            {
                scoreBuilder.SetPageLayout(_pageLayoutParser.Parse(pageLayoutElement));
            }

            var systemLayoutElement = defaultsElement.Element("system-layout");
            if (systemLayoutElement != null)
            {
                scoreBuilder.SetDefaultSystemLayout(_systemLayoutParser.Parse(systemLayoutElement));
            }

            var staffLayouts = new List<StaffLayout>();
            foreach (var staffLayoutElement in defaultsElement.Elements("staff-layout"))
            {
                staffLayouts.Add(_staffLayoutParser.Parse(staffLayoutElement));
            }
            if (staffLayouts.Count > 0)
            {
                scoreBuilder.setDefaultStaffLayouts(staffLayouts);
            }

            var appearanceElement = defaultsElement.Element("appearance");
            if (appearanceElement != null)
            {
                scoreBuilder.SetAppearance(ParseAppearance(appearanceElement));
            }
        }

        private Credit ParseCredit(XElement creditElement)
        {
            var page = XmlHelper.GetAttributeValueAsInt(creditElement, "page");
            var creditType = XmlHelper.GetElementText(creditElement, "credit-type");
            var creditWords = new List<string>();
            
            foreach (var wordsElement in creditElement.Elements("credit-words"))
            {
                var text = wordsElement.Value.Trim();
                if (!string.IsNullOrEmpty(text))
                {
                    creditWords.Add(text);
                }
            }

            return new Credit(page, creditType, creditWords);
        }

        private List<Part> ParsePartList(XElement partListElement)
        {
            var parts = new List<Part>();

            // 获取整个文档的 <part> 节点集合
            var doc = partListElement.Document;
            var partNodes = doc?.Descendants("part").ToList();

            foreach (var child in partListElement.Elements())
            {
                switch (child.Name.LocalName)
                {
                    case "score-part":
                        var partId = child.Attribute("id")?.Value;
                        var partNode = partNodes?.FirstOrDefault(p => p.Attribute("id")?.Value == partId);
                        if (partNode != null)
                        {
                            // 调试输出
                            Console.WriteLine($"[ScoreParser] <part id={partId}> measure count: {partNode.Elements("measure").Count()}");
                            var part = _partParser.Parse(partNode, partListElement);
                            if (part != null)
                            {
                                parts.Add(part);
                            }
                        }
                        else
                        {
                            Console.WriteLine($"[ScoreParser] <part id={partId}> not found in document");
                        }
                        break;
                    case "part-group":
                        // 处理part-group逻辑
                        WarningSystem.AddWarning(
                            message: "Part-group elements are not yet fully supported",
                            category: WarningCategories.Generic,
                            rule: "part_group_not_supported",
                            line: XmlHelper.GetLineNumber(child),
                            elementName: "part-group",
                            context: CreateContext()
                        );
                        break;
                }
            }

            return parts;
        }

        private Appearance ParseAppearance(XElement appearanceElement)
        {
            var lineWidths = new List<LineWidth>();
            var noteSizes = new List<NoteSize>();

            foreach (var child in appearanceElement.Elements())
            {
                switch (child.Name.LocalName)
                {
                    case "line-width":
                        lineWidths.Add(ParseLineWidth(child));
                        break;
                    case "note-size":
                        noteSizes.Add(ParseNoteSize(child));
                        break;
                }
            }

            return new Appearance(lineWidths, noteSizes);
        }

        private LineWidth ParseLineWidth(XElement element)
        {
            var type = XmlHelper.GetAttributeValue(element, "type");
            var width = XmlHelper.GetElementTextAsDouble(element) ?? 0.0;

            return new LineWidth(type, width);
        }

        private NoteSize ParseNoteSize(XElement element)
        {
            var type = XmlHelper.GetAttributeValue(element, "type");
            var size = XmlHelper.GetElementTextAsDouble(element) ?? 0.0;

            return new NoteSize(type, size);
        }
    }
}
